fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/AvailableFields.java

```diff
@@ -0,0 +1,211 @@
+                                                : commonPrimaryKey.normalizeKeyForPositions();
+        index.trimPrimaryKey(primaryKeys);
+        keyFields.addAll(primaryKeys);
+
+        Map<KeyExpression, FieldData> fields = new HashMap<>();
+        final IndexKeyValueToPartialRecord.Builder builder = IndexKeyValueToPartialRecord.newBuilder(recordType);
+        for (int i = 0; i < keyFields.size(); i++) {
+            KeyExpression keyField = keyFields.get(i);
+            FieldData fieldData = FieldData.of(IndexKeyValueToPartialRecord.TupleSource.KEY, i);
+            if (!keyField.createsDuplicates() && addCoveringField(keyField, fieldData, builder)) {
+                fields.put(keyField, fieldData);
+            }
+        }
+        for (int i = 0; i < valueFields.size(); i++) {
+            KeyExpression valueField = valueFields.get(i);
+            FieldData fieldData = FieldData.of(IndexKeyValueToPartialRecord.TupleSource.VALUE, i);
+            if (!valueField.createsDuplicates() && addCoveringField(valueField, fieldData, builder)) {
+                fields.put(valueField, fieldData);
+            }
+        }
+
+        if (!builder.isValid()) {
+            return NO_FIELDS;
+        }
+        return new AvailableFields(fields);
+    }
+
+    public static boolean addCoveringField(@Nonnull KeyExpression requiredExpr,
+                                           @Nonnull FieldData fieldData,
+                                           @Nonnull IndexKeyValueToPartialRecord.Builder builder) {
+        while (requiredExpr instanceof NestingKeyExpression) {
+            NestingKeyExpression nesting = (NestingKeyExpression)requiredExpr;
+            String fieldName = nesting.getParent().getFieldName();
+            requiredExpr = nesting.getChild();
+            builder = builder.getFieldBuilder(fieldName);
+        }
+        if (requiredExpr instanceof FieldKeyExpression) {
+            FieldKeyExpression fieldKeyExpression = (FieldKeyExpression)requiredExpr;
+            if (!fieldKeyExpression.getNullStandin().equals(Key.Evaluated.NullStandin.NULL) ||
+                    fieldKeyExpression.getFanType().equals(KeyExpression.FanType.FanOut)) {
+                return false;
+            }
+            builder.addField(fieldKeyExpression.getFieldName(), fieldData.source, fieldData.index);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Nonnull
+    public static AvailableFields intersection(@Nonnull List<AvailableFields> toIntersect) {
+        Map<KeyExpression, FieldData> intersection = null;
+        for (AvailableFields fields : toIntersect) {
+            if (!fields.hasAllFields()) {
+                if (intersection == null) {
+                    intersection = new HashMap<>(fields.fields);
+                } else {
+                    Set<KeyExpression> keysToRemove = new HashSet<>(Sets.difference(intersection.keySet(),
```

Comment:
`Map.keySet` is backed by the map itself, so I think `retainAll` should work here.

Reply:
Ah, great, that makes this much tidier.